### PR TITLE
sidecar: trim service ports based on egress listeners and merge services

### PR DESCRIPTION
**Pull request by ramaraochavali**
This PR does two things for sidecar
1. When Egress listener specifies a port, it trims other ports of the service
2. When each egress listener produces a partial service, it merges services across listeners in the sidecar scope
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@rshriram PTAL, implemented as discussed.

File: pilot/pkg/model/service.go [hunk 1]
```diff
@@ -653,3 +654,38 @@ func (s *Service) GetServiceAddressForProxy(node *Proxy) string {
 	}
 	return s.Address
 }
+
+// DeepCopy creates a clone of Service.
+// TODO : See if there is any efficient alternative to this function - copystructure can not used as is because
+// Service has sync.RWMutex that can not be copied. One alternative is to convert that to a pointer but that needs
+// lot of changes.
```
Comment (rshriram) [inline pilot/pkg/model/service.go]: we dont need to. the sync mutex is used by multi-cluster code. I think it should be pushed out @hzxuzhonghu

File: pilot/pkg/model/sidecar.go [hunk 2]
```diff
@@ -211,7 +209,7 @@ func ConvertToSidecarScope(ps *PushContext, sidecarConfig *Config, configNamespa
 	// Now collect all the imported services across all egress listeners in
 	// this sidecar crd. This is needed to generate CDS output
 	out.services = make([]*Service, 0)
-	servicesAdded := make(map[string]struct{})
```
Comment (rshriram) [inline pilot/pkg/model/sidecar.go]: actually, per some blog, this is more memory efficient than bool :P.

Comment (rshriram) [reply pilot/pkg/model/sidecar.go]: How about we make this map of string and service pointer (same pointer in the array?) This way, in the if/else block below, we can cut down the iteration, and simply merge with the existing one in the map, while maintaining the order in the array.

File: pilot/pkg/model/sidecar.go [hunk 3]
```diff
@@ -477,20 +482,36 @@ func (ilw *IstioEgressListenerWrapper) selectServices(services []*Service, confi
...
 					} else {
-						portMatched = true
+						importedServices = append(importedServices, s)
+						hostFound = true
+						break
 					}
+					// If there is a port match, we should trim the service ports to the port specified by listener.
 					if portMatched {
-						importedServices = append(importedServices, s)
+						sc := s.DeepCopy()
```
Comment (rshriram) [inline pilot/pkg/model/sidecar.go]: how about we move the deep copy inside the if block? [also add a break statement inside the if block]

File: pilot/pkg/model/sidecar.go (hunk 2 above)
Comment (ramaraochavali) [reply pilot/pkg/model/sidecar.go]: > actually, per some blog, this is more memory efficient than bool :P.

Interesting.. Thanks for pointing that. I will change to service pointer

File: pilot/pkg/model/service.go (hunk 1 above)
Comment (ramaraochavali) [reply pilot/pkg/model/service.go]: If it is pushed out of Service, we can directly use `copystructure`

**Comment by ramaraochavali** [comment]
@rshriram changed as suggested, PTAL

**Pull request merged**